controllers/teleport_controller/teleport_controller.cpp

```diff
@@ -0,0 +1,138 @@
+// File:          teleport_controller.cpp
+// Date:          23/03/2021
+// Description:   Method to teleport robot around the playing field for the purposes of
+//                    gathering training data
+// Author:        Lachlan Court
+// Modifications:
+
+
+#include <RobotisOp2GaitManager.hpp>
+#include <RobotisOp2MotionManager.hpp>
+#include <webots/Supervisor.hpp>
+#include <iostream>
+#include <math.h>
+#include <time.h> 
+
+// All the webots classes are defined in the "webots" namespace
+using namespace webots;
+
+/*
+// Rotations matrix for 90 degree rotations
+static const int rotationsSize = 4;
+static const double rotations[rotationsSize][4] = {{1, 0, 0, 1.57081},
+                                                  {0.577353, 0.577345, 0.577353, 2.0944},
+                                                  {1.69526e-09, 0.707103, 0.707111, 3.14159},
+                                                  {-0.577352, 0.577347, 0.577352, -2.09441}};
+
+*/
+
+// Rotations matrix for 45 degree rotations
+static const int rotationsSize = 8;                                     
+static const double rotations[rotationsSize][4] = {{1, 0, 0, 1.57081},
```

Comment:
There are a few things I would change about this approach, and other things I would tell you for future reference. In no particular order:
- these constants are known at compile time, so we could (if we were going to keep them here (we're probably not going to)) make them `constexpr`'s instead of just `const`. `constexpr` is an even nicer version of const, but it has to be knowable at compile time.
- We don't usually use arrays like this. `std::array` is nicer, if we want to use arrays, but that's quite rare. More often we would use `std::vector`, because dynamic sizing is nice (these are basically C++'s version of java's arraylists). In this case, in the main codebase we would probably use a `std::array` of one of eigen's vector types (our linear algebra library of choice), since the number of rotations is constant. I am not sure if this repo will have eigen any time soon though, so again, it'll be down to people who are working on such things to guide you here.
- It's not clear to me what these constants are meant to be doing. The comment "Rotations matrix for 45 degree rotations" doesn't make sense to me in this context, because this is far too many constants to be such a matrix.
- If it is a rotation matrix, it is probably named [against our conventions](https://nubook.nubots.net/system/foundations/mathematics#rotations). These are confusing at first, but intuitive once you understand them. @ysims  can help you with naming matrices ;~)
- The biggest thing, is that these constants probably shouldn't even be in this file. They should ideally be read from a config/data file, like in the `.yaml` file [in this directory](https://github.com/NUbots/NUbots/tree/master/module/support/NUsight/data/config). Convert these into a yaml file, then read them from there.

Reply:
The rotations in webots are in a format I've never seen before, 4 values, x y z range from -1 to 1 to indicate an axis of rotation that the 4th value, the angle should be applied to. Perhaps rotation matrix was the wrong term to use but it made sense to me, essentially I manually moved the robot around in the GUI to determine the rotations I wanted to use and copied them in here because it was too hard to calculate them on the fly (I'm sure it is possible but not being familiar with this method of handling rotation I was not confident delving into the maths of it all)

I am not sure how to use yaml files so I might need a hand with that but if that's the route i should be taking more than happy to look into it

Reply:
sounds like they might using [unit quaternions](https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation) for rotation, but without naming it that? (I don't know enough about quaternions for rotations...) @ysims @Bidski is this a unit quaternion set up? [relevant docs page](https://cyberbotics.com/doc/reference/transform)

Reply:
Calling them matrices is definitely not appropriate though IMO

Reply:
Sounds like [AxisAngle](https://en.m.wikipedia.org/wiki/Axis%E2%80%93angle_representation) tbh.

Reply:
Ahh true. I thought that was an eigen-only thing lol. that looks like exactly what we're looking for

Reply:
Yeah I think axis angle is the one, it also has an option to view the rotation in quarternions but I don't know how to set it as quarternions

Reply:
Axis angle and unit quaternions are quite similar given the fact that they are all from the range -1 to 1. Axis angle the angle part would often be greater than or less than -1 to 1. An easy way to check is to see if the 2 norm of the values is 1.
If we are using Eigen in here and this is one of our constants, you should be storing these in one of Eigen's types (AngleAxisd or Quaterniond at least if not Affine3d`

Reply:
I've never heard of those before aha I'll confirm which one it is and swap it over. Given that there are a number of different rotations, should I be storing each of those Eigen types in an std::vector? Am I understanding that correctly?

Reply:
Yea, make an `std::vector<Eigen::AngleAxisd>` or a `std::array<Eigen::AngleAxisd, N>` depending on the context (where `N` is the number of elements in the array, known at compile-time)

Reply:
If you know how many there are and it won't change, use `std::array`, if you will be adding/removing elements to it then use `std::vector`

Reply:
Gotcha, probs do an array and I can have a size in the yaml config file

Reply:
If your size is coming from a yaml file it's dynamic because you don't know it without running the code

Reply:
I think I have misunderstood how a yaml file works then, I thought I would be able to read the size, create the array according to that size, and then read the rotations and add them into the array?

Reply:
To use an `std::array` you need to know exactly how many elements will be stored in it at compile-time. If the size of the array is stored in a YAML file then you wont know the size of the array until run-time, so you will have to use a `std::vector` instead.

Reply:
there's nothing special about `yaml` files. That's just the format we use for config files - it's basically formatted plaintext. The way we (usually) read the data from them is by reading them at run time, with NUClear emitting a `Configuration` message. The module reading the `yaml` gets a config object it can use, in an `on<Configuration>(..)` call. It uses that config object to assign the config values to member variables of the module. [Random example from the codebase](https://github.com/NUbots/NUbots/blob/master/module/motion/QuinticWalk/src/QuinticWalk.cpp#L68).

In this case, since it's not a module (or more accurately, a NUClear reactor), we don't have that formal configuration process with the `on` statement. Instead, you'll probably want to use the library which does it for us directly: [yaml-cpp](https://github.com/jbeder/yaml-cpp). 

A big benefit to doing this generally is that we can tweak the parameters to change how the program behaves without going deep into the trenches to find what we should change. We won't end up wondering if we're changing the right thing, or if that change will fundamentally break the program (hopefully). A similar effect can be realised by naming magic numbers generally (although this case isn't quite a magic number, but more a set of magic tuples), but consistently using config/data files makes it easy.